contracts/README.md

```diff
@@ -3,15 +3,396 @@
+- Must implement the `IRegistry` interface for standard resolution
+- All registries store data in the singleton `RegistryDatastore`
+
+**Root Registry** → **TLD Registries** (.eth, .box, etc.) → **Domain Registries** (example.eth) → **Subdomain Registries** (sub.example.eth)
+
+**Resolution Process**
+1. Start at root registry
+2. Recursively traverse to find the deepest registry with a resolver set
+3. Query that resolver for the requested record
+4. Supports wildcard resolution (parent resolver handles subdomains)
+
+### Token ID System
+
+ENSv2 uses a dual-identifier system to enable token regeneration while maintaining stable storage:
+
+#### How Token IDs Work
+
+```solidity
+// Label to Token ID conversion
+string label = "example";
+bytes32 labelHash = keccak256(bytes(label));
+uint256 tokenId = uint256(labelHash);
+
+// Canonical ID derivation (XOR with lower 32 bits)
+uint256 canonicalId = tokenId ^ uint32(tokenId);
+```
+
+#### Key Properties
+
+1. **Token ID** (`uint256`): Unique ERC1155 identifier that can change over time
+   - Different tokenIds can exist for the same name (e.g., after expiry/re-registration)
+   - Lower 32 bits may encode version, timestamp, or other metadata
+
+2. **Canonical ID** (`uint256`): Stable storage key that never changes
+   - Always has lower 32 bits zeroed out
+   - Formula: `canonicalId = tokenId ^ uint32(tokenId)`
+   - Idempotent: `getCanonicalId(getCanonicalId(x)) == getCanonicalId(x)`
+
+#### Why This Matters
+
+```solidity
+// Example: Name lifecycle
+// First registration
+uint256 tokenId1 = 0x9c22ff5f21f0b81b113e63f7db6da94fedef11b2119b4088b89664fb9a3cb658;
+uint256 canonical = 0x9c22ff5f21f0b81b113e63f7db6da94fedef11b200000000000000000000000;
+
+// After expiry and re-registration (different tokenId!)
+uint256 tokenId2 = 0x9c22ff5f21f0b81b113e63f7db6da94fedef11b212345678901234567890abcd;
+uint256 canonical = 0x9c22ff5f21f0b81b113e63f7db6da94fedef11b200000000000000000000000; // SAME!
+
+// All storage lookups use canonical ID
+datastore.getEntry(registry, canonical); // Always finds the current data
+```
+
+**Enables**:
+- **Token Regeneration**: Names can be burned and re-minted without losing storage
+- **Version/Timestamp Encoding**: Lower 32 bits available for metadata
+- **Consistent Storage Access**: Single storage slot lookup regardless of tokenId variations
+- **No Storage Duplication**: Different tokenIds for same name share storage
+
+#### Usage in Code
+
+```solidity
+import {LibLabel} from "./utils/LibLabel.sol";
+
+// Convert label to canonical ID
+uint256 canonicalId = LibLabel.labelToCanonicalId("example");
+
+// Get canonical version of any ID
+uint256 canonical = LibLabel.getCanonicalId(someTokenId);
+
+// Storage always uses canonical IDs
+DATASTORE.setSubregistry(canonicalId, subregistryAddr);
+```
+
+### Access Control
+
+ENSv2 uses **Enhanced Access Control (EAC)**, a bitmap-based permission system that stores up to 32 roles and their assignee counts in a single uint256:
```

Comment:
replace "permission" with "role-based access control".

Upto 32 roles are allowed, with each role have a complementary _admin_ role that grants the holder permission to grant/revoke that role. Roles are assigned against a "resource", which is just a `uint256` that can represent anything. For a given role assigned for a given resource, a maximum of 15 assignees are allowed. For registered names the canonical token ID acts as the equivalent access control resource id.